alacritty_terminal/src/selection.rs

```diff
@@ -40,35 +40,39 @@ use crate::term::{Search, Term};
 /// [`semantic`]: enum.Selection.html#method.semantic
 /// [`lines`]: enum.Selection.html#method.lines
 #[derive(Debug, Clone, PartialEq)]
+pub enum SelectionType {
+    Simple,
+    Semantic,
+    Lines,
+}
+
+/// Describes a region of a 2-dimensional area
+/// 
+/// Used to track the selection. The selection type can be heterogeneous
+/// where each end of the region can be a different selection type.
+#[derive(Debug, Clone, PartialEq)]
 pub enum Selection {
-    Simple {
+    Normal {
         /// The region representing start and end of cursor movement
         region: Range<Anchor>,
     },
     Block {
         /// The region representing start and end of cursor movement
         region: Range<Anchor>,
     },
-    Semantic {
-        /// The region representing start and end of cursor movement
-        region: Range<Point<isize>>,
-    },
-    Lines {
-        /// The region representing start and end of cursor movement
-        region: Range<Point<isize>>,
-    },
 }
 
 /// A Point and side within that point.
 #[derive(Debug, Clone, PartialEq)]
 pub struct Anchor {
-    point: Point<isize>,
+    pub point: Point<isize>,
+    pub ty: SelectionType,
```

Comment:
I'm sorry if I'm missing something, but why do the anchors need to store the `SelectionType`? As I understand it, a selection is defined by a starting point and an ending point. Editing a selection only depends on the starting point, and depending on the mouse clicks will update the end point according to one of the three selections types.

Storing the `SelectionType` could be useful if we were trying to preserve the type while modifying the selection, but this isn't the case. I don't think this is a functionality issue, but more of a data structures concern/question.

Seems related to https://github.com/jwilm/alacritty/pull/2361/files#r294029075

Thoughts?

Reply:
> why do the anchors need to store the `SelectionType`?

I currently have `SelectionType` stored in the anchors because `to_span` needs to know how to expand the visual selection. I moved it into the `Anchor` because it explains how the `Anchor` should be expanded and that makes more sense stored in the `Anchor` than in `Selection`.

> Editing a selection only depends on the starting point, and depending on the mouse clicks will update the end point according to one of the three selections types.

There is one situation where editing (updating) a selection will have an effect on both ends of the selection. This is where the user expands the selection in the same mouse click that created a semantic or line expansion. In this case if the user changes the direction that they drag the selection then the semantic/line selection start point needs to expand in the opposite direction than what was initially done.

On another note, perhaps a better name for the enum would be `ExpansionType` instead of `SelectionType`.